go/libraries/doltcore/sqle/kvexec/merge_join.go

```diff
@@ -0,0 +1,498 @@
+	// left key. |l.nextRightKey| can be a lookahead key or nil at the
+	// end of this stage.
+	l.nextRightKey, l.nextRightVal, err = l.rightIter.Next(ctx)
+	if err != nil {
+		if errors.Is(err, io.EOF) {
+			// this is OK, but need to skip nil key comparison
+			goto match
+		}
+		return nil, err
+	}
+	if l.lrCmp(l.leftKey, l.leftVal, l.nextRightKey, l.nextRightVal) == 0 {
+		l.lookaheadBuf = append(l.lookaheadBuf, l.nextRightKey, l.nextRightVal)
+		goto matchBuf
+	}
+
+match:
+	// We start this stage with at least one match. We consider the
+	// lookahead buffer matches before |l.rightKey| because we can
+	// track state with a single |l.matchPos| variable. Matching merge
+	// condition does not guarantee the rest of the filters will match.
+	if l.matchPos < len(l.lookaheadBuf) {
+		candidate, ok, err := l.tryReturn(ctx, l.leftKey, l.leftVal, l.lookaheadBuf[l.matchPos], l.lookaheadBuf[l.matchPos+1])
+		if err != nil {
+			return nil, err
+		}
+		l.matchPos += 2
+		l.matchedLeft = ok
+		if !ok {
+			goto match
+		}
+		return candidate, nil
+	} else if l.matchPos == len(l.lookaheadBuf) {
+		candidate, ok, err := l.tryReturn(ctx, l.leftKey, l.leftVal, l.rightKey, l.rightVal)
+		if err != nil {
+			return nil, err
+		}
+		l.matchedLeft = ok
+		l.matchPos++
+		if !ok {
+			goto match
+		}
+		return candidate, nil
+	} else {
+		// We exhausted matches for the current |l.leftKey|.
+		// See whether we should flush or reuse the right buffer
+		// with the next left key.
+		l.matchPos = 0
+
+		// compare the current to the next left key
+		tmpKey, tmpVal := l.leftKey, l.leftVal
+		l.leftKey, l.leftVal, err = l.leftIter.Next(ctx)
+		if err != nil {
+			return nil, err
+		}
+		cmp := l.llCmp(tmpKey, tmpVal, l.leftKey, l.leftVal)
+
+		if cmp != 0 {
+			// if the left key is new, invalidate lookahead buffer and
+			// advance the right side
+			l.lookaheadBuf = l.lookaheadBuf[:0]
+			if l.nextRightKey != nil {
+				l.rightKey, l.rightVal = l.nextRightKey, l.nextRightVal
+				l.nextRightKey = nil
+			} else {
+				l.rightKey, l.rightVal, err = l.rightIter.Next(ctx)
+				if err != nil {
+					if errors.Is(err, io.EOF) && l.isLeftJoin {
+						// need to check |tmpKey| for left-join before
+						// jumping to exhaustLeft
+						l.exhaustLeft = true
+					} else {
+						return nil, err
+					}
+				}
+			}
+		}
+
+		if l.isLeftJoin && !l.matchedLeft {
+			// consider left join after appropriate state transitions
+			ret, ok, err := l.tryReturn(ctx, tmpKey, tmpVal, nil, nil)
+			if err != nil {
+				return nil, err
+			}
+			if ok {
+				return ret, nil
+			}
+		}
+		l.matchedLeft = false
+
+		if cmp == 0 {
+			// The left keys are equivalent, the right-side lookahead
+			// buffer is still valid for the new left key. The only reason
+			// we didn't short-circuit this check earlier is for left-joins.
+			goto match
+		}
+		if l.exhaustLeft {
+			goto exhaustLeft
+		}
+		goto incr
+	}
+
+exhaustLeft:
```

Comment:
Again, more naturally expressed as a loop

Reply:
Also I think you can get away with making this a method instead of a label, and everywhere you `goto exhaustLeft` just `return l.exhaustLeft()`